src/sentry/snuba/entity_subscription.py

```diff
@@ -0,0 +1,274 @@
+
+    if query:
+        return f"({event_type_conditions}) AND ({query})"
+
+    return event_type_conditions
+
+
+@dataclass
+class _EntitySubscription:
+    entity_key: str
+    dataset: str
+    time_col: str
+
+
+class BaseEntitySubscription(ABC, _EntitySubscription):
+    """
+    An abstraction layer for all different entity subscriptions. It is important to note that
+    this abstraction layer was added because the subscription logic was too coupled to the
+    events and transactions entities, which was fine initially but now as we are adding more
+    entities to support subscriptions (alerts), we need to decouple this logic.
+    """
+
+    def __init__(self, aggregate: str, extra_fields: Optional[Mapping[str, Any]] = None):
+        self.time_col = ENTITY_TIME_COLUMNS[self.entity_key]
+
+    @abstractmethod
+    def build_snuba_filter(
+        self,
+        query: str,
+        environment: Optional[Environment],
+        params: Optional[Mapping[str, Any]] = None,
+    ) -> Filter:
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_entity_extra_params(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+
+class BaseEventsAndTransactionEntitySubscription(BaseEntitySubscription, ABC):
+    def __init__(self, aggregate: str, extra_fields: Optional[Mapping[str, Any]] = None):
+        super().__init__(aggregate, extra_fields)
+        self.aggregate = aggregate
+        self.event_types = None
+        if extra_fields:
+            self.event_types = extra_fields.get("event_types")
+
+    def build_snuba_filter(
+        self,
+        query: str,
+        environment: Optional[Environment],
+        params: Optional[Mapping[str, Any]] = None,
+    ) -> Filter:
+        resolve_func = resolve_column(Dataset(self.dataset))
+
+        query = apply_dataset_query_conditions(QueryDatasets(self.dataset), query, self.event_types)
+        snuba_filter = get_filter(query, params=params)
+        snuba_filter.update_with(
+            resolve_field_list([self.aggregate], snuba_filter, auto_fields=False)
+        )
+        snuba_filter = resolve_snuba_aliases(snuba_filter, resolve_func)[0]
+        if snuba_filter.group_ids:
+            snuba_filter.conditions.append(
+                ["group_id", "IN", list(map(int, snuba_filter.group_ids))]
+            )
+        if environment:
+            snuba_filter.conditions.append(["environment", "=", environment.name])
+        return snuba_filter
+
+    def get_entity_extra_params(self) -> Mapping[str, Any]:
+        return {}
+
+
+class EventsEntitySubscription(BaseEventsAndTransactionEntitySubscription):
+    dataset = Dataset.Events.value
+    entity_key = EntityKey.Events.value
+
+
+class TransactionsEntitySubscription(BaseEventsAndTransactionEntitySubscription):
+    dataset = Dataset.Transactions.value
+    entity_key = EntityKey.Transactions.value
+
+
+class SessionsEntitySubscription(BaseEntitySubscription):
+    dataset = Dataset.Sessions.value
```

Comment:
Why do we use the value vs the Enum?

Reply:
I believe I started this in `sentry.release_health` because I didn't know how enums worked and it propagated everywhere.

Reply:
I will look into this tomorrow again.. I don't remember tbh but I remember there was a reason why I opted for using the string here